File: HS.jl
using LinearAlgebra

function spin(a,n)
   s=zeros(Int8,n)
    for i=1:n
        s[i]=(rem(a,2^i)-rem(a,2^(i-1)))/2^(i-1)
    end
    return s
end


function hs(n)
    h=zeros(2^n,2^n)
    for a=0:2^n-1
        for b=0:2^n-1
            spa=spin(a,n)
            spb=spin(b,n)

            for i=2:n
                for j=1:i-1
                    if spa[i]==spb[i]&& spa[j]==spb[j]
                        h[a+1,b+1]=h[a+1,b+1]+1/(4*sin(pi*(i-j)/n)^2);
                    elseif spa[i]==spb[j]&& spa[j]==spb[i]
                        h[a+1,b+1]=h[a+1,b+1]-1/(4*sin(pi*(i-j)/n)^2);
                    end
                end
            end
        end
    end
    return h
end

function hse(n)
    return eigvals(hs(n))
end

File: HS.py
import numpy as np
from scipy.linalg import eigh

def spin(a, n):
    s = np.zeros(n, dtype=np.int8)
    for i in range(n):
        s[i] = (a % (2 ** (i+1)) - a % (2 ** i )) // (2 ** i)
    return s

def hs(n):
    h = np.zeros((2 ** n, 2 ** n))
    for a in range(2 ** n):
        for b in range(2 ** n):
            spa = spin(a, n)
            spb = spin(b, n)

            for i in range(1, n):
                for j in range(i):
                    if spa[i] == spb[i] and spa[j] == spb[j]:
                        h[a, b] += 1 / (4 * np.sin(np.pi * (i - j) / n) ** 2)
                    elif spa[i] == spb[j] and spa[j] == spb[i]:
                        h[a, b] -= 1 / (4 * np.sin(np.pi * (i - j) / n) ** 2)
    return h

def hse(n):
    return eigh(hs(n))[0]

